extensions/amp-story-shopping/0.1/amp-story-shopping-config.js

```diff
@@ -14,44 +12,63 @@ import {
  *  items: !Array<!ShoppingConfigDataDef>,
  * }}
  */
-let ShoppingConfigDef;
-
-export class AmpStoryShoppingConfig extends AMP.BaseElement {
-  /** @param {!AmpElement} element */
-  constructor(element) {
-    super(element);
-    /** @private @const {?../../amp-story/1.0/amp-story-store-service.AmpStoryStoreService} */
-    this.storeService_ = null;
+let ShoppingConfigResponseDef;
+
+/** @typedef {!Object<string, !ShoppingConfigDataDef> */
+export let KeyedShoppingConfigDef;
+
+/** @type {!WeakMap<!Element, !Promise<!KeyedShoppingConfigDef>>} */
+const cache = new WeakMap();
```

Comment:
The fact that we need a layer of caching on top of the existing layer of caching (store service) looks like a design flaw. Is there any reason why the shopping data is not stored in the store service by page ID? @processprocess

Reply:
The cache layer is optional. The store service holds the data regardless. 

I'm adding it because the store service currently does not scope items per page, and allows ids to be overridden. If that behavior changes, the cache becomes moot.

Reply:
Sounds like the store service should do just that, you're bringing up a second good reason why it should :))

Reply:
Sounds good. I removed the cache, and added a TODO for page-level scoping.

Reply:
It's not stored by page ID so that it can be used on different pages.
In this way publishers do not need to repeat data if they want to feature the same product on multiple pages.

Reply:
What's the configuration to repeat a product across another page?
If I understand correctly, config is required for all pages so there's no way to skip fetching the config to re-use a product data.

I don't feel strongly either way but I felt like calling out a pattern that should have been unnecessary and came from this design choice, as this is how code debt is built.

Reply:
> What's the configuration to repeat a product across another page?  

Use the same `product-tag-id`. As a product on a different page. 
This breaks though since data is only loaded when the page is close to the viewport.

> config is required for all pages so there's no way to skip fetching the config to re-use a product data.

Good point.
If product data is configured on page 10, but the tag is on page 1 it will break.
The only way to share data and ensure it's loaded is to load all data at once.
It definitely needs to be refactored to ensure it always works.

I'll move this discussion to the design doc.